File: src/main/webapp/resources/js/reportController.js
/**
 * Created by kivi on 18.12.17.
 */

function ReportService($resource) {
    return $resource('rest/report/:id', {id: '@id'});
}

function ReportPermService($resource) {
    return $resource('rest/report/:id/permissions?user=:login', {id: '@id', login: '@login'});
}

function ReportProjectService($resource) {
    return $resource('rest/report/:id/project', {id: '@id'});
}

function ReportCommentService($resource) {
    return $resource('rest/report/:id/comments?user=:login', {id: '@id', login: '@login'});
}

function ReportController($scope, $http, $routeParams,
                          ReportService,
                          ReportPermService,
                          ReportCommentService,
                          ReportProjectService,
                          InfoShareService) {
    function url() {
        return {id: $routeParams.id};
    }
    function url_with_login(login) {
        return {id: $routeParams.id, login:login};
    }

    this.user = InfoShareService.getUser();
    this.permissions = ReportPermService.get(url_with_login(this.user.login));
    this.project = ReportProjectService.get(url());
    this.instance = ReportService.get(url());
    this.comments = ReportCommentService.query(url());

    this.changeStatus = function (status) {
        $http.put('rest/report/' + this.instance.id + '?user=' + this.user.login, status)
            .then(function () {
                this.update();
                this.updateComments();
            }.bind(this), function (error) {
                alert(error.data.message);
        });
    };

    this.commentReport = function () {
        if (isEmpty($scope.reportComment)) {
            alert("Enter comment message");
        } else {
            var comment = new ReportCommentService();
            comment.description = $scope.reportComment;
            comment.$save(url_with_login(this.user.login), function () {
                $scope.reportComment = "";
                this.updateComments();
            }.bind(this), function (error) {
                alert(error.data.message);
            });
        }
    };

    this.update = function () {
        this.instance = ReportService.get(url());
    };

    this.updateComments = function () {
        this.comments = ReportCommentService.query(url());
    }
}

app
    .factory('ReportService', ReportService)
    .factory('ReportPermService', ReportPermService)
    .factory('ReportProjectService', ReportProjectService)
    .factory('ReportCommentService', ReportCommentService)
    .controller('ReportController', ReportController);

File: src/main/java/com/kspt/pms/controller/ReportController.java
package com.kspt.pms.controller;

import com.kspt.pms.entity.BugReport;
import com.kspt.pms.entity.Comment;
import com.kspt.pms.entity.Project;
import com.kspt.pms.entity.User;
import com.kspt.pms.exception.NotFoundException;
import com.kspt.pms.logic.Developer;
import com.kspt.pms.logic.Permissions;
import com.kspt.pms.logic.Tester;
import com.kspt.pms.repository.BugReportRepository;
import com.kspt.pms.repository.CommentRepository;
import com.kspt.pms.repository.MessageRepository;
import com.kspt.pms.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.Collection;

/**
 * Created by kivi on 18.12.17.
 */
@RestController
public class ReportController {
    @Autowired
    BugReportRepository bugReportRepository;
    @Autowired
    UserRepository userRepository;
    @Autowired
    CommentRepository commentRepository;
    @Autowired
    MessageRepository messageRepository;

    @RequestMapping("/rest/report/{id}")
    public BugReport getReport(@PathVariable Long id) {
        return bugReportRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Report " + id.toString()));
    }

    @RequestMapping("/rest/report/{id}/permissions")
    public Permissions getPermissions(@PathVariable Long id,
                                      @RequestParam("user") String login) {
        BugReport report = bugReportRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Report " + id.toString()));
        User user = userRepository.findByLogin(login)
                .orElseThrow(() -> new NotFoundException(login));
        return Permissions.getPermissionsByRole(report.getProject().getRoleForUser(user));
    }

    @RequestMapping("/rest/report/{id}/project")
    public Project getProject(@PathVariable Long id) {
        BugReport report = bugReportRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Report " + id.toString()));
        return report.getProject();
    }

    @RequestMapping(value = "/rest/report/{id}", method = RequestMethod.PUT)
    public void updateStatus(@PathVariable Long id,
                             @RequestParam("user") String login,
                             @RequestBody String status) {
        BugReport report = bugReportRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Report " + id.toString()));
        User user = userRepository.findByLogin(login)
                .orElseThrow(() -> new NotFoundException(login));
        Tester tester = new Tester(user, bugReportRepository, commentRepository, messageRepository);
        Developer developer = new Developer(user, bugReportRepository, commentRepository, messageRepository);
        switch (status) {
            case "OPENED":
                tester.reopenReport(report);
                break;
            case "ACCEPTED":
                developer.acceptReport(report);
                break;
            case "FIXED":
                developer.fixReport(report);
                break;
            case "CLOSED":
                tester.closeReport(report);
                break;
            default: break;
        }
        bugReportRepository.save(report);
    }

    @RequestMapping("/rest/report/{id}/comments")
    public Collection<Comment> getComments(@PathVariable Long id) {
        BugReport report = bugReportRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Report " + id.toString()));
        return report.getComments();
    }

    @RequestMapping(value = "/rest/report/{id}/comments", method = RequestMethod.POST)
    public void getComments(@PathVariable Long id,
                            @RequestParam("user") String login,
                            @RequestBody Comment comment) {
        BugReport report = bugReportRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Report " + id.toString()));
        User user = userRepository.findByLogin(login)
                .orElseThrow(() -> new NotFoundException(login));
        Developer developer = new Developer(user, bugReportRepository, commentRepository, messageRepository);
        developer.commentReport(report, comment.getDescription());
        bugReportRepository.save(report);
    }
}
